cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -386,125 +408,91 @@ func StartControllers(s *options.CMServer, kubeconfig *restclient.Config, rootCl
-		}
+	if availableResources[unversioned.GroupVersionResource{Group: "policy", Version: "v1beta1", Resource: "poddisruptionbudgets"}] {
+		glog.Infof("Starting disruption controller")
+		go disruption.NewDisruptionController(sharedInformers.Pods().Informer(), clientBuilder.ClientOrDie("disruption-controller")).Run(stop)
+		time.Sleep(wait.Jitter(s.ControllerStartInterval.Duration, ControllerStartJitter))
 	}
 
-	groupVersion = "apps/v1beta1"
-	resources, found = resourceMap[groupVersion]
-	glog.Infof("Attempting to start statefulset, full resource map %+v", resourceMap)
-	if containsVersion(versions, groupVersion) && found {
-		glog.Infof("Starting %s apis", groupVersion)
-		if containsResource(resources, "statefulsets") {
-			glog.Infof("Starting StatefulSet controller")
-			resyncPeriod := ResyncPeriod(s)()
-			go petset.NewStatefulSetController(
-				sharedInformers.Pods().Informer(),
-				client("statefulset-controller"),
-				resyncPeriod,
-			).Run(1, wait.NeverStop)
-			time.Sleep(wait.Jitter(s.ControllerStartInterval.Duration, ControllerStartJitter))
-		}
+	if availableResources[unversioned.GroupVersionResource{Group: "apps", Version: "v1beta1", Resource: "statefulsets"}] {
+		glog.Infof("Starting StatefulSet controller")
+		resyncPeriod := ResyncPeriod(s)()
+		go petset.NewStatefulSetController(
+			sharedInformers.Pods().Informer(),
+			clientBuilder.ClientOrDie("statefulset-controller"),
+			resyncPeriod,
+		).Run(1, stop)
+		time.Sleep(wait.Jitter(s.ControllerStartInterval.Duration, ControllerStartJitter))
 	}
 
-	groupVersion = "batch/v2alpha1"
-	resources, found = resourceMap[groupVersion]
-	if containsVersion(versions, groupVersion) && found {
-		glog.Infof("Starting %s apis", groupVersion)
-		if containsResource(resources, "cronjobs") {
-			glog.Infof("Starting cronjob controller")
-			// // TODO: this is a temp fix for allowing kubeClient list v2alpha1 sj, should switch to using clientset
-			kubeconfig.ContentConfig.GroupVersion = &unversioned.GroupVersion{Group: batch.GroupName, Version: "v2alpha1"}
-			go cronjob.NewCronJobController(client("cronjob-controller")).
-				Run(wait.NeverStop)
-			time.Sleep(wait.Jitter(s.ControllerStartInterval.Duration, ControllerStartJitter))
-			time.Sleep(wait.Jitter(s.ControllerStartInterval.Duration, ControllerStartJitter))
-		}
-	} else {
-		glog.Infof("Not starting %s apis", groupVersion)
+	if availableResources[unversioned.GroupVersionResource{Group: "batch", Version: "v2alpha1", Resource: "cronjobs"}] {
+		glog.Infof("Starting cronjob controller")
+		// TODO: this is a temp fix for allowing kubeClient list v2alpha1 sj, should switch to using clientset
+		cronjobConfig := rootClientBuilder.ConfigOrDie("cronjob-controller")
+		cronjobConfig.ContentConfig.GroupVersion = &unversioned.GroupVersion{Group: batch.GroupName, Version: "v2alpha1"}
+		go cronjob.NewCronJobController(clientset.NewForConfigOrDie(cronjobConfig)).Run(stop)
```

Comment:
Regarding the user-agent comment: only the `Builder.Client` func sets it, not `Builder.Config`. I guess we have to use the rootClientBuilder here therefore, not `clientset.NewForConfigOrDie`.